### ISSUE TITLE
Question about reserved names

### ISSUE THREAD

**Issue by sungam3r** [graphql/graphql-spec#827]
See https://spec.graphql.org/June2018/#sec-Reserved-Names :

> Types and fields required by the GraphQL introspection system that are used in the same context as user‐defined types and fields are prefixed with "__" two underscores. This in order to avoid naming collisions with user‐defined GraphQL types. Conversely, GraphQL type system authors must not define any types, fields, arguments, **or any other type system artifact** with two leading underscores.

The question is - is enum **value** such _any other type system artifact_? In other words, is it possible to have `__TEST` enum value?

Relates to https://github.com/graphql-dotnet/graphql-dotnet/pull/2380

**Comment by sungam3r** [comment]
@benjie Can you help with this please?

**Comment by benjie** [comment]
An enum is a type system artifact, so `__TEST` is not allowed. I've confirmed that the reference implementation forbids this:

```js
const { GraphQLSchema, GraphQLEnumType, GraphQLObjectType, validateSchema, printSchema } = require('graphql');

const MyEnum = new GraphQLEnumType({
  name: "MyEnum",
  values: {
    __A: {
      value: 'A'
    }
  }
});

const Query = new GraphQLObjectType({
  name: 'Query',
  fields: {
    a: {
      type: MyEnum
    }
  }
});

const schema = new GraphQLSchema({query: Query});

const errors = validateSchema(schema)
if (errors && errors.length) {
  console.error(errors);
  throw new Error("INVALID SCHEMA");
}

console.log(printSchema(schema));
```

```
[
  GraphQLError [Object]: Name "__A" must not begin with "__", which is reserved by GraphQL introspection.
      at isValidNameError (/home/benjie/Dev/graphile/postgraphile/node_modules/graphql/utilities/assertValidName.js:38:12)
```

**Comment by sungam3r** [comment]
Thanks

**Comment by leebyron** [comment]
What the reference implementation does is assert that any `Name` cannot start with `__` unless it's being specially defined for introspection. I think the spec text you've referenced is not particularly clear about this, and this important bit is fairly buried - I've included a hopeful clarification to this in #857
